### PR TITLE
[WIP] Implement T3-rtx timer and Congestion Control

### LINKED ISSUE #11: Add T3-rtx timer and congestion control
https://tools.ietf.org/html/rfc4960#section-6.3.3

### PR DESCRIPTION
**Pull request by enobufs**
Basic implemention is done. Working on adding more tests right now.

Open issue: how to report T3-rtx retranmission failure to ULP?

See my [implementation note](https://docs.google.com/document/d/18HxKodWkPPfZpKTddQQvepux5MOS3QWMfK1eJYXQdz4) for your review

Resolves #11

### CODE REVIEW

**Comment by backkem** [comment]
Sidenote: It would be super awesome if we can capture some of your implementation notes on the pion.ly knowledge base.

File: association.go [hunk 1]
```diff
@@ -1070,21 +1471,22 @@ func (a *Association) handleChunk(p *packet, c chunk) ([]*packet, error) {
 		return nil, nil
 
 	case *chunkAbort:
-		fmt.Println("Abort chunk, with errors:")
+		a.log.Debugf("Abort chunk, with follwing errors:")
```
Comment (backkem) [inline association.go]:
```suggestion
		a.log.Debugf("Abort chunk, with following errors:")
```

File: payload_queue.go [hunk 2]
```diff
@@ -1,86 +1,153 @@
...
 
-type payloadDataArray []*chunkPayloadData
+////////////////////////////////////////////////////////////////////////////////
+// Pending Queue
 
-func (s payloadDataArray) search(tsn uint32) (*chunkPayloadData, bool) {
-	i := sort.Search(len(s), func(i int) bool {
-		return sna32GTE(s[i].tsn, tsn)
-	})
+type pendingQueue struct {
```
Comment (backkem) [inline payload_queue.go]: Maybe just move this to it's own file instead?

File: association.go (hunk 1 above)
Comment (enobufs) [reply association.go]: Fixed in 472b620

File: payload_queue.go (hunk 2 above)
Comment (enobufs) [reply payload_queue.go]: Agreed and fixed in 472b620

**Comment by Antonito** [comment]
Very nice work! 
I tried it with https://github.com/Antonito/gfile, I managed to successfully transfer a 500MB file between my computer and a small VPS at a constant speed of 400-600kB/s.
`Network: 524280000 bytes | 19m56.054232771s | 0.4180 MB/s`
It's still slow, but way better than before.

On my loopback, I achieved ~20MB/s, I still have to try over local network.

Also the results seem to be the best when I use a 512kB threshold.

**Comment by enobufs** [comment]
Waiting for #30 to land first.

File: association.go [hunk 3]
```diff
@@ -277,45 +302,56 @@ func (a *Association) sendCookieEcho() error {
 
 // Close ends the SCTP Association and cleans up any state
 func (a *Association) Close() error {
+	a.log.Debugf("[%s] closing association..", a.name())
+
+	// TODO: shutdown sequence
+	a.lock.Lock()
```
Comment (Sean-Der) [inline association.go]: Is it possible/worth making `setState` an atomic, the less locking the better (IMO)

File: association.go [hunk 4]
```diff
@@ -277,45 +302,56 @@ func (a *Association) sendCookieEcho() error {
...
+	a.log.Debugf("[%s] association closed", a.name())
 	return nil
 }
 
 func (a *Association) readLoop() {
+	var closeErr error
 	defer func() {
 		a.lock.Lock()
-		closeErr := errors.New("association closed")
+		//closeErr := errors.New("association closed")
```
Comment (Sean-Der) [inline association.go]: Should we delete this line (or uncomment)?

File: association.go [hunk 5]
```diff
@@ -1200,23 +1688,41 @@ func (a *Association) onRetransmissionFailure(id int) {
...
+
+	return a.pendingQueue.getNumBytes() + a.inflightQueue.getNumBytes()
 }
 
-/*
 func (a *Association) name() string {
-	if a.netConn != nil && a.netConn.LocalAddr() != nil {
-		return a.netConn.LocalAddr().String()
-	}
+	/*
```
Comment (Sean-Der) [inline association.go]: Delete (or bring back) uncommented code

File: association.go [hunk 6]
```diff
@@ -519,20 +579,73 @@ func (a *Association) handleInitAck(p *packet, i *chunkInitAck) error {
...
+		}},
+	})
+}
+
+func (a *Association) handleCookieEcho(c *chunkCookieEcho) []*packet {
+	if !bytes.Equal(a.myCookie.cookie, c.cookie) {
+		return nil
+	}
+
+	// stop T1-init timer
```
Comment (kc5nra) [inline association.go]: These comments seem a bit redundant as the code is really clear

File: association.go [hunk 7]
```diff
@@ -924,24 +1280,136 @@ func (a *Association) resetStreams() *packet {
...
+	a.inflightQueue.pushNoCheck(c)
+}
+
+// popPendingDataChunksToSend pops chunks from the pending queues as many as
+// the cwnd and rwnd allows to send.
+// The caller should hold the lock.
+func (a *Association) popPendingDataChunksToSend() []*chunkPayloadData {
+	chunks := []*chunkPayloadData{}
+
+	if a.pendingQueue.size() > 0 {
```
Comment (kc5nra) [inline association.go]: We check this several times, should we just make an empty()?

File: association.go [hunk 8]
```diff
@@ -200,11 +220,20 @@ func createAssocation(netConn net.Conn) *Association {
...
+	//  o  The initial cwnd before DATA transmission or after a sufficiently
+	//     long idle period MUST be set to min(4*MTU, max (2*MTU, 4380
+	//     bytes)).
+	a.cwnd = min32(4*a.mtu, max32(2*a.mtu, 4380))
+	a.log.Tracef("updated cwnd=%d ssthresh=%d inflight=%d (INI)", a.cwnd, a.ssthresh, a.inflightQueue.getNumBytes())
+
 	a.t1Init = newRTXTimer(timerT1Init, a, maxInitRetrans)
 	a.t1Cookie = newRTXTimer(timerT1Cookie, a, maxInitRetrans)
-	a.t3RTX = newRTXTimer(timerT3RTX, a, pathMaxRetrans)
+	a.t3RTX = newRTXTimer(timerT3RTX, a, 0 /* rtx forever */)
```
Comment (kc5nra) [inline association.go]: Perhaps 0 is a constant rather than a comment? Not 100% sure on this one

Comment (enobufs) [reply association.go]: Good idea.

File: association.go (hunk 5 above)
Comment (enobufs) [reply association.go]: will delete.

File: association.go (hunk 4 above)
Comment (enobufs) [reply association.go]: will delete it.

File: association.go (hunk 7 above)
Comment (enobufs) [reply association.go]: size() is used in the tests (payload_queue_test.go) a lot. I think adding empty() would also check the length of the internal queue (no performance gain), and size() has more "information" than empty().  Let me keep those lines...

File: association.go (hunk 3 above)
Comment (enobufs) [reply association.go]:
```
func (a *Association) setState(state associationState) {
	if a.state != state {
		a.log.Debugf("[%s] state change: '%s' => '%s'", a.name(), a.state.String(), state.String())
		a.state = state
	}
}
```
I found logging before and after state is useful.  Caller is assume to hold the mutext in the current code, so not much gain.  I will add a comment to say, caller must hold the lock instead!

**Comment by Antonito** [comment]
The latests updates really improve the performances.
I managed to do some file exchanges via [gfile](https://github.com/Antonito/gfile/tree/develop):
- 500MB at `16.8301 MB/s` on my loopback (with peaks at `29MB/s`)
- 50MB at `0.8207 MB/s` between my computer (near Los Angeles) and a small VPC I own in Paris,  with peaks at `1.20MB/s`.
For reference, I manage to transfer files at `~4MB/s` via `scp` (`ssh`).

I tried increasing the receiving window to 1MB, it led to a minimal performance increase (I managed to exchange data at `1.5MB/s` with the VPC,  almost no difference on the loopback). 
It seems the main bottleneck for my app is now the packet loss. :D

See CPU benchmarks:
[cpu-sender-50MB-VPC.pdf](https://github.com/pion/sctp/files/3046085/cpu-sender-50MB-VPC.pdf)
[cpu-sender-500MB-loopback.pdf](https://github.com/pion/sctp/files/3046087/cpu-sender-500MB-loopback.pdf)

**Comment by enobufs** [comment]
Thanks so much @Antonio for sharing the profiling results. Looks like the improvement of queues (sort only when necessary) is working well too.  As anticipated, readLoop()->handleIncoming() now stands out.  I believe we can improve that by introducing Delayed Ack (#32) next.

File: association.go (hunk 4 above)
Comment (enobufs) [reply association.go]: Fixed at ef7368e

File: association.go (hunk 6 above)
Comment (enobufs) [reply association.go]: Fixed at ef7368e

File: association.go (hunk 8 above)
Comment (enobufs) [reply association.go]: Fixed at ef7368e

File: association.go (hunk 5 above)
Comment (enobufs) [reply association.go]: Fixed at ef7368e

File: association.go (hunk 3 above)
Comment (enobufs) [reply association.go]: Fixed at ef7368e

File: association.go [hunk 9]
```diff
@@ -630,42 +752,262 @@ func (a *Association) getOrCreateStream(streamIdentifier uint16) *Stream {
...
+	// HTNA algorithm - RFC 4960 Sec 7.2.4
+	// Increment missIndicator of each chunks that the SACK reported missing
+	// when either of the following is met:
+	// a)  Not in fast-recovery, or;
+	// b)  In fast-recovery AND the Cumulative TSN Ack Point advanced
+	if !a.inFastRecovery || (a.inFastRecovery && cumTSNAckPointAdvanced) {
+		for tsn := cumTSNAckPoint + 1; sna32LTE(tsn, htna); tsn++ {
+			c, ok := a.inflightQueue.get(tsn)
+			if !ok {
+				return nil, errors.Errorf("Requested non-existent TSN %v", tsn)
```
Comment (Antonito) [inline association.go]: Capitalized error message, `errors.Errorf` vs `fmt.Errorf`

File: association.go [hunk 10]
```diff
@@ -630,42 +752,262 @@ func (a *Association) getOrCreateStream(streamIdentifier uint16) *Stream {
...
 
 	// New ack point, so pop all ACKed packets from inflightQueue
 	// We add 1 because the "currentAckPoint" has already been popped from the inflight queue
 	// For the first SACK we take care of this by setting the ackpoint to cumAck - 1
 	for i := a.cumulativeTSNAckPoint + 1; sna32LTE(i, d.cumulativeTSNAck); i++ {
-		if _, ok := a.inflightQueue.pop(i); !ok {
-			return nil, errors.Errorf("TSN %v unable to be popped from inflight queue", i)
+		c, ok := a.inflightQueue.pop(i)
+		if !ok {
+			return 0, 0, errors.Errorf("TSN %v unable to be popped from inflight queue", i)
```
Comment (Antonito) [inline association.go]: `errors.Errorf` vs `fmt.Errorf`

File: association.go [hunk 11]
```diff
@@ -155,13 +173,13 @@ func Server(netConn net.Conn) (*Association, error) {
 		}
 		return a, nil
 	case <-a.closeCh:
-		return nil, errors.Errorf("Assocation closed before connecting")
+		return nil, errors.Errorf("Association closed before connecting")
```
[the remark this answers is not in the document]
Comment (enobufs) [reply association.go]: Fixed at af5449e

File: association.go (hunk 10 above)
Comment (enobufs) [reply association.go]: Fixed at af5449e

File: association.go [hunk 12]
```diff
@@ -172,26 +190,28 @@ func Client(netConn net.Conn) (*Association, error) {
...
```
[the remark this answers is not in the document]
Comment (enobufs) [reply association.go]: Fixed at 602f899

File: association.go [hunk 13]
```diff
@@ -630,42 +752,262 @@ func (a *Association) getOrCreateStream(streamIdentifier uint16) *Stream {
...
 
 	// Mark selectively acknowledged chunks as "acked"
 	for _, g := range d.gapAckBlocks {
 		for i := g.start; i <= g.end; i++ {
-			c, ok := a.inflightQueue.get(d.cumulativeTSNAck + uint32(i))
+			tsn := d.cumulativeTSNAck + uint32(i)
+			c, ok := a.inflightQueue.get(tsn)
 			if !ok {
-				return nil, errors.Errorf("Requested non-existent TSN %v", d.cumulativeTSNAck+uint32(i))
+				return 0, 0, errors.Errorf("Requested non-existent TSN %v", tsn)
```
[the remark this answers is not in the document]
Comment (enobufs) [reply association.go]: Fixed at 602f899

File: association.go (hunk 9 above)
Comment (enobufs) [reply association.go]: Fixed at 602f899

File: association.go [hunk 14]
```diff
@@ -952,73 +1427,102 @@ func (a *Association) sendPayloadData(chunks []*chunkPayloadData) error {
...
 
-	return packets
+	packets := a.bundleDataChunksIntoPackets(chunks)
+
+	a.lock.Unlock()
+	var err error
+
+	for _, p := range packets {
+		if err = a.send(p); err != nil {
+			err = errors.Wrap(err, "Unable to send outbound packet")
```
[the remark this answers is not in the document]
Comment (enobufs) [reply association.go]: Fixed at 602f899

**Pull request merged**